README.md

|  Version  | Docs  | Status |
| :--- 	  | :--- 	  | :--- 	  |
|  [![PyPi Version](https://img.shields.io/pypi/v/pynastran.svg)](https://pypi.python.org/pypi/pyNastran) | [docs](https://pynastran-git.readthedocs.io/en/1.2/) | [![Build Status](https://img.shields.io/travis/SteveDoyle2/pyNastran/1.2.svg)](https://travis-ci.org/SteveDoyle2/pyNastran) [![Coverage Status](https://img.shields.io/coveralls/SteveDoyle2/pyNastran/1.2.svg)](https://coveralls.io/github/SteveDoyle2/pyNastran?branch=1.2) |
|   Master | [![Documentation Status](https://readthedocs.org/projects/pynastran-git/badge/?version=latest)](http://pynastran-git.readthedocs.io/en/latest/?badge=latest) | [![Linux Status](https://img.shields.io/travis/SteveDoyle2/pyNastran/master.svg)](https://travis-ci.org/SteveDoyle2/pyNastran) ![Coverage Status](https://coveralls.io/repos/github/SteveDoyle2/pyNastran/badge.svg?branch=master) |



<!---
[![Windows Status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi/branch/master?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran)
[![codecov](https://codecov.io/gh/SteveDoyle2/pyNastran/branch/master/graph/badge.svg)](https://codecov.io/gh/SteveDoyle2/pyNastran)
[![Coverage Status](https://img.shields.io/coveralls/SteveDoyle2/pyNastran/master.svg)](https://coveralls.io/github/SteveDoyle2/pyNastran?branch=master)
--->

<!---
[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi/branch/master?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran/branch/master)
[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran)
[![Build Status](https://travis-ci.org/SteveDoyle2/pyNastran.png)](https://travis-ci.org/SteveDoyle2/pyNastran)
[![Coverage Status](https://coveralls.io/repos/github/SteveDoyle2/pyNastran/badge.svg?branch=master)](https://coveralls.io/github/SteveDoyle2/pyNastran?branch=master)
[![Requirements Status](https://img.shields.io/requires/github/SteveDoyle2/pyNastran/master.svg)](https://requires.io/github/SteveDoyle2/pyNastran/requirements/?branch=master)	|
--->

### v1.2.1

[Download GUI](https://sourceforge.net/projects/pynastran/files/?source=navbar) (latest is from 2019/5/24)

Also, check out the:
  * [Discussion forum](http://groups.google.com/group/pynastran-discuss) (intended for questions about the latest release)
  * [Developer forum](http://groups.google.com/group/pynastran-dev) (intended for questions about the master branch)
  * [Docs](https://pynastran-git.readthedocs.io/en/1.2/)

for more detailed information.

# Overview

pyNastran is an interface library to the various Nastran file formats (BDF, OP2, OP4).
Using the BDF interface, you can read/edit/write Nastran geometry without worrying about
field formatting.  Many checks are also performed to verify that your model is correct.
Using the OP2 interface, you can read large result files quckly and efficiently.
Additionally, you can also extract a subset of the result data and write OP2/F06 result
files.  For a more detailed list of features, see:
  * [Features](https://pynastran-git.readthedocs.io/en/1.2/quick_start/features.html#overview)

Using the pyNastran GUI, you can read in Nastran models and quickly view results for a model.
While it's no FEMAP/Patran, it can replace many tasks that would otherwise require a
commercial program.

![GUI](https://github.com/SteveDoyle2/pynastran/blob/master/pyNastran/gui/images/caero.png)


# News

### pyNastran v1.2.1 has been released (2019/5/24)

I keep saying to myself there's not much to add, but Nastran is huge.  Beyond HDF5
support in the BDF, I'm a huge fan of the new ability to keep track of which include file a
card came from and write it as a separate file.  It's limited in usefulness, but very handy
in certain cases.  There's a new (still preliminary) superelement capability.  I'm far
from a superelement expert, but it's probably useful.

The OP2 reader now supports SORT2 along with much improved random results reading.
If you're using 60+ GB OP2s, you probably have had issues with RAM usage in the past.
With the new ability to dump the OP2 directly to HDF5, this should not be an as much of
an issue.  It's not 100% implemented, so let me know if you need it for another result.

Regarding the GUI, there are also some new features.  Groups work a bit better, but aren't
quite perfect.  Logging has been dramatically sped up so the GUI loads faster and you can
load Nastran models even faster if you disable additional results (e.g., element quality).

Finally, Python 2.7 is end of life.  Numpy, scipy, and matplotlib have all dropped
Python 2.7 support.  It's time for pyNastran to as well.  The OP2 reader is 30% faster in
Python 3.6+ than Python 2.7, so it's not all bad!

Programmatics:
 - This is the last Python 2.7 release.
 - Dropping Python 3.5 support
 - Supports Python 2.7, 3.6-3.7
 - GUI is compatible with PyQt4/PyQt5 as well as PySide/PySide2
 - improved testing of old versions of packages

BDF:
 - 373 cards supported (up from 343)
 - added abiltity to write models to separate include files
     ```python
     >>> model = BDF()
     >>> model.read_bdf(bdf_filename, save_file_structure=True)

     out_filenames = {
         bdf_filename : bdf_filename_new,
         include_filename : include_filename_new,
     }
     >>> model.write_bdfs(out_filenames, relative_dirname=None, is_windows=None)
     >>> ifile = model.grids[1].ifile
     ```

 - HDF5 import/export
      ```python
     >>> model = read_bdf(bdf_filename)
     >>> model.export_hdf5_filename(hdf5_filename)
     >>> model_new = OP2()
     >>> model_new.load_hdf5_filename(hdf5_filename)
     ```

 - preliminary superelement support
     ```python
     >>> model.read_bdf(bdf_filename)
     >>> model.superelement_models[1].nodes
     ```


OP2:
 - reorganization of random op2 results into op2.results.psd (or ato, no, crm, rms) to aide in finding data
 - reorganization of op2 class to reduce number of functions in the object.  This affects any custom table reading.
 - improved optimzation response reading
 - limited SORT2 support
 - fixed CD transformation bug for BOUGV1 and BOPHIG1 tables
 - Improved HDF5 export/import support (e.g., matrices, random results)

 - Can optionally save directly to HDF5 instead of numpy (limited).
 - Loading OP2s to an HDF5 file to decrease memory usage
      ```python
     >>> op2_model = OP2()
     >>> op2_model.load_as_h5 = True
     >>> op2_model.read_op2(op2_filename)
     ```

OP2Geom:
 - HDF5 support
 - reading EQEXIN/S, GPT, GPDT, CSTM/S tables (recovery of nodes & coordinate with OP2Geom)
 - fixed theta/mcid reading for CTRIA3/CQUAD4
 - fixed CQUAD8 bug

GUI:
 - sped up HTML logging
 - much improved groups menu
 - options for Nastran in preferences menu to speed up loading/limit memory usage
 - pyNastran BDF pickle reading
 - pyNastran OP2 HDF5 reading (not MSC's format)
 - visualization when pickling nodes/elements
 - min/max labels
 - highlight menu
 - Patran-style colors
 - custom force vectors
 - AVL support


Known issues:
 - Transient Pandas Dataframes will fail for newer versions of numpy/pandas.
   If anyone knows how to use a MultiIndex, this is probably pretty easy to fix.

### pyNastran v1.2.0 has been released (2019/5/21)

This result has been superseeded by 1.2.1.  See release notes for details.

### pyNastran v1.1.0 has been released (2018/6/26)

|  Version  | Docs  | Status |
| :--- 	  | :--- 	  | :--- 	  |
|  [![PyPi Version](https://img.shields.io/pypi/v/pynastran.svg)](https://pypi.python.org/pypi/pyNastran) | [docs](http://pynastran.m4-engineering.com/1.1.0/) | [![Build Status](https://img.shields.io/travis/SteveDoyle2/pyNastran/v1.1.svg)](https://travis-ci.org/SteveDoyle2/pyNastran) [![Coverage Status](https://img.shields.io/coveralls/SteveDoyle2/pyNastran/v1.1.svg)](https://coveralls.io/github/SteveDoyle2/pyNastran?branch=v1.1) |

It's been roughly a year and ~100 tickets closed since the last version, so it's probably
time for another release!  I want to thank everybody who helped with this release, especially
Michael Redmond.  He is working on h5Nastran, which which ties in with pyNastran.  It's not quite
ready yet, but it offers the possibility of major speedups for large models.


Probably the most best thing is the documentation actually builds again.  It was too big
and took too long, so readthedocs failed.  I'm pleased to announce that the continuing
problems of up-to-date documentation will hopefully be a thing of the past.  [M4 Engineering](http://www.m4-engineering.com)
has offered to host the documentation on http://www.pynastran.m4-engineering.com.
There's still some work to do regarding hosting documentation for older versions and the master,
but that's hopefully coming soon.  Outside of that, it's the same open-source project
and will still be on Github.

Regarding features, the focus has again been on robustness and testing.  There has been
a 10% increase in the testing coverage (the same as v0.8 to v1.0).  There are a few
changes (mainly in the BDF) though.  The GUI now also supports PyQt4, PyQt5, and Pyside
with the same API, so it's a bit easier to install from source as simplifying licensing
issues as PyQt is GPL.

Programmatics:
 - Dropping Python 3.4 support
 - Supports Python 2.7, 3.5-3.6
 - dropping VTK 5/6 suppoprt for the GUI

BDF:
 - 343 cards supported (up from 312)
 - cross-referencing is now more straightforward to new users (much of v1.0 works using the `_ref` option)
   - ``*_ref`` attributes are cross-referenced
      - ``element.nodes`` is not cross-referenced
      - ``element.nodes_ref`` is cross-referenced
 - pickling to reload your deck ~5x faster
 - decreased time required for Case Control Deck with large SETs and many load cases
 - improved optimization checks

OP2:
 - HDF5 export/import support
      ```python
     >>> op2_model = read_op2(op2_filename)
     >>> op2_model.export_hdf5_filename(hdf5_filename)
     >>> op2_model_new = OP2()
     >>> op2_model_new.load_hdf5_filename(hdf5_filename, combine=True)
     ```
 - pandas support for matrices
 - couple more results vectorized (e.g., complex strain energy, DMIG strain energy, some forces)
 - grid_point_stressses supported (disabled since v0.7)
 - fixed sparse matrices being stored as dense matrices

GUI:
 - preliminary support for PySide
 - can now mix and match fringe/displacement/vector results (e.g., max principal stress shown on a displaced model)
 - improved animation menu
    - in gui animation
    - more animation profiles
 - bar profile visualzation
 - nominal geometry (useful for deflection plots)
 - improved optimization support
 - improved picking display
 - better PSHELL/PCOMP distinction

Known issues:
 - Transient Pandas Dataframes will fail for newer versions of numpy/pandas.
   If anyone knows how to use a MultiIndex, this is probably pretty easy to fix.

### pyNastran v1.0.0 has been released (2017/5/25)

See [v1.0.0](https://github.com/SteveDoyle2/pyNastran/releases/tag/v1.0.0) for information regarding enhancements.


### pyNastran v0.8.0 has been released (2016/8/21)

|  Version  | Docs  |
| :--- 	  | :--- 	  |
|  [v0.8.0](https://github.com/SteveDoyle2/pyNastran/releases/tag/v0.8.0) | [![Documentation Status](https://readthedocs.org/projects/pynastran-git/badge/?version=v0.8)](http://pynastran-git.readthedocs.io/en/v0.8.0/?badge=v0.8.0) |

<!---
[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi/branch/master?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran/branch/master)
[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran)
--->

See [v0.8.0](https://github.com/SteveDoyle2/pyNastran/releases/tag/v0.8.0) for information regarding enhancements.


### pyNastran v0.7.2 has been Released (2015/4/25)

See [v0.7.2](https://github.com/SteveDoyle2/pyNastran/releases) for information regarding enhancements.

### Version 0.6.1 has been released (2013/6)
**Version 0.6** improves BDF reading.  The reader is more robust and also requires proper BDF field formatting (e.g. a integer field can't be a float).  Additionally, cards also have a comment() method.

Marcin Gąsiorek participated in the latest pyNastran under the European Space Agency's (ESA) "Summer of Code In Space" [SOCIS](http://sophia.estec.esa.int/socis2012/?q=node/13) program.  The program provides a stipend to students to work on open-source projects.
He did a great job of simplifying code and creating nicer documentation.


releaseNotes.txt
This document discusses the major features (and limitations) added in each
version of the software.  See the HTML documentation for more information.

If you like it/want to help leave some feedback on the
Dicussion Page:       https://groups.google.com/forum/#!forum/pynastran-discuss
Developer Dicussion:  https://groups.google.com/forum/#!forum/pynastran-dev

If you have a bug/want a new feature or card, leave some feedback on the
Issues List:     https://github.com/SteveDoyle2/pyNastran/issues

*** Not done....

-----------------------v1.2.2 released 2020/1/x-----------------------
BDF:
 - bug fixes:
   - PBARL & PBEAML DBOX now has defaults for DIM5-DIM10
   - added AECOMPL, DEFORM to bdf attributes
   - added RHO to MAT10 updater
   - Better case control SET parser (can now handle 1,THRU,10,EXCEPT,7,8,
     which doesn't have spaces that makes parsing easier).  It's also orders
     of magnitude faster on large problems.
   - Fixed NDDL bug in MONPNT2
   - Added check for DVPRELx to prevent referencing property fields when another
     similar property is used (e.g., a DVPRELx references a PCOMPG, but the PCOMP
     field id numbers are used to update the card).
   - added some missing hdf5 object support
   - fixing DEQATN bug caused by using python builtin
   - fixed missing variable in mass properties error message
   - aero
     - fixed issue where CAERO1 didn't set the xyz1 and xyz4 points to arrays
       when loading an HDF5 file
     - caero2 now sets box ids
   - elements/properties
     - added I12, J to PBEAM updater; fixed error message
     - fixed Area calculation for PBARL/CHAN, HAT1, DBOX
     - fixed PBAR optimization treating I12 as I1; added J
   - convert better handles:
     - T1-T4 in quad shells
     - fixed bug in CGAP x/g0
   - bdf mirror:
     - fixed bug where max element id is less than max rigid_element id
     - fixed bug with shells that have blanks (e.g., CTRIA6, CQUAD8)
     - plane is now case insensitive and auto-sorted
     - CAERO1 now supports lspan/lchord and Cp
     - CAERO1 handles xy plane properly now (doesn't assume it's xz)
     - fixing bug in set_param
OP2:
 - geom bug fixes:
    - fixed reading bugs for QHBDY and SPCOFF
    - better identifing duplicate property id (PBCOMP -> PBEAM)
    - fixed CONVM error (it can be 6 or 7 fields, not only 6)
    - fixed ACCEL key
 - bug fixes:
    - R1TABRG now stores response_types as strings (instead of bytes for
      Python 3)
    - f06 writer now writes CDAMP3/4 names correctly

GUI:
 - bug fixes:
  - added check on highlight menu for model existance
  - fixed import for new version download menu
  - improved command line error message
  - fixed support for CAEROx models without elements
  - export_cases now supports integers

-----------------------v1.2.1 released 2019/5/24-----------------------
OP2:
 - fixed bug with OUGV1PAT table

-----------------------v1.2.0 released 2019/5/21-----------------------
Programmatics:
 - This is the last Python 2.7 release.
 - Dropping Python 3.5 support
 - Supports Python 2.7, 3.6-3.7
 - GUI is compatible with PyQt4/PyQt5 as well as PySide/PySide2
 - improved testing of old versions of packages

BDF:
 - 373 cards supported (up from 343)
 - added abiltity to write models to separate include files
     ```python
     >>> model = BDF()
     >>> model.read_bdf(bdf_filename, save_file_structure=True)

     out_filenames = {
         bdf_filename : bdf_filename_new,
         include_filename : include_filename_new,
     }
     >>> model.write_bdfs(out_filenames, relative_dirname=None, is_windows=None)
     >>> ifile = model.grids[1].ifile
     ```

 - HDF5 import/export
      ```python
     >>> model = read_bdf(bdf_filename)
     >>> model.export_hdf5_filename(hdf5_filename)
     >>> model_new = OP2()
     >>> model_new.load_hdf5_filename(hdf5_filename)
     ```

 - preliminary superelement support
     ```python
     >>> model.read_bdf(bdf_filename)
     >>> model.superelement_models[1].nodes
     ```


OP2:
 - reorganization of random op2 results into op2.results.psd (or ato, no, crm, rms) to aide in finding data
 - reorganization of op2 class to reduce number of functions in the object.  This affects any custom table reading.
 - improved optimzation response reading
 - limited SORT2 support
 - fixed CD transformation bug for BOUGV1 and BOPHIG1 tables
 - Improved HDF5 export/import support (e.g., matrices, random results)

 - Can optionally save directly to HDF5 instead of numpy (limited).
 - Loading OP2s to an HDF5 file to decrease memory usage
      ```python
     >>> op2_model = OP2()
     >>> op2_model.load_as_h5 = True
     >>> op2_model.read_op2(op2_filename)
     ```

OP2Geom:
 - HDF5 support
 - reading EQEXIN/S, GPT, GPDT, CSTM/S tables (recovery of nodes & coordinate with OP2Geom)
 - fixed theta/mcid reading for CTRIA3/CQUAD4
 - fixed CQUAD8 bug

GUI:
 - sped up HTML logging
 - much improved groups menu
 - options for Nastran in preferences menu to speed up loading/limit memory usage
 - pyNastran BDF pickle reading
 - pyNastran OP2 HDF5 reading (not MSC's format)
 - visualization when pickling nodes/elements
 - min/max labels
 - highlight menu
 - Patran-style colors
 - custom force vectors
 - AVL support


Known issues:
 - Transient Pandas Dataframes will fail for newer versions of numpy/pandas.  If anyone knows how to use a MultiIndex,
   this is probably pretty easy to fix.

-----------------------v1.1.0 released 2018/6/26-----------------------
This is a major release.  As in the v1.0 release, the focus has been on testing.

The bigest change has been to cross-referencing.  All cross-referenced
objects now have an ``_ref`` attribute that is cross-referened, which
is ``None`` if the model is not cross referenced.  Objects maintain the
uncross-referenced form as well.  For a CQUAD4, instead of:
 ``model.elements[eid].nodes[0].nid``
the referenced object is:
 ``model.elements[eid].nodes_ref[0].nid``


packages.py
"""helper for setup.py"""
import os
import sys

PY2 = False
PY3 = True
if sys.version_info < (3, 0):
    PY2 = True
    PY3 = False
PY36 = sys.version_info[:2] == (3, 6)
#PY37 = sys.version_info[:2] == (3, 7)

# features in packages used by pyNastran
# numpy
#  - 1.12 min for 3.6
#  - 1.13: adds axis support to unique
#  - 1.14: adds encoding support to savetxt (unused)
#  - 1.14: adds proper writing of np.savetxt for open file objects
#          (used for unicode savetxt using with statement) in Python 3.6
#  - 1.15: min for Python 3.7? I guess 1.14 is fine for a requirement...
# scipy:
#  - 0.18.1: fixed kdtree used by nodal equivalencing; min for Python 2.7
#  - 0.19:   min for Python 3.6
#  - 0.19:   min for Python 3.7?; last 0.x release
# matplotlib:
#  - 1.5: min for Python 2.7; last 1.x release
#  - 2.0: adds C0/C1 colors (use colors from default colormap);
#    min for Python 3.6
#  - 2.1: adds plt.subplots support (untested?)
#  - 2.2: min for Python 3.7
# pandas:
#  - 0.20, <0.25 - we'll check if the user has pandas installed (also at runtime)

# the packages that change requirements based on python version
REQS = {
    '2.7' : {
        # package : (less than check, required)
        'numpy' : ('1.13', '>=1.13,<1.17'),
        'scipy' : ('0.18.1', '>=0.18.1,<1.13'),
        'matplotlib' : ('1.5', '>=1.5,<3'),
    },
    '3.6' : {
        'numpy' : ('1.14', '>=1.14'),
        'scipy' : ('0.19', '>=0.19'),
        'matplotlib' : ('2.0', '>=2.0'),
    },
    '3.7' : {
        'numpy' : ('1.14', '>=1.14'),
        'scipy' : ('1.0', '>=1.0'),
        'matplotlib' : ('2.2', '>=2.2'),  # 2.2.4 adds Python 3.7 support
    },
}

def check_python_version():
    """verifies the python version"""
    imajor, minor1, minor2 = sys.version_info[:3]
    if sys.version_info < (2, 7, 7):  # 2.7.15 used
        # makes sure we don't get the following bug:
        #   Issue #19099: The struct module now supports Unicode format strings.
        sys.exit('Upgrade your Python to >= 2.7.7 or 3.6+; version=(%s.%s.%s)' % (
            imajor, minor1, minor2))

    if PY3:
        if sys.version_info < (3, 6, 0):  # 3.7.1 used
            sys.exit('Upgrade your Python to >= 2.7.7 or 3.6+; version=(%s.%s.%s)' % (
                imajor, minor1, minor2))


def int_version(name, version):
    """splits the version into a tuple of integers"""
    sversion = version.split('-')[0]
    #numpy
    #scipy
    #matplotlib
    #qtpy
    #vtk
    #cpylog
    #pyNastran
    if 'rc' not in name:
        # it's gotta be something...
        # matplotlib3.1rc1
        sversion = sversion.split('rc')[0]

    try:
        return [int(val) for val in sversion.split('.')]
    except ValueError:
        raise SyntaxError('cannot determine version for %s %s' % (name, sversion))


def str_version(version):
    """converts a tuple of intergers to a version number"""
    return '.'.join(str(versioni) for versioni in version)


def get_package_requirements(is_gui=True, add_vtk_qt=True, python_version=None):
    """gets the requirements for setup.py"""
    if python_version is None:
        python_version = '%s.%s' % sys.version_info[:2]

    if python_version not in REQS:
        python_version = '3.7'
    vreqs = REQS[python_version]

    all_reqs = {}

    is_dev = (
        'TRAVIS' in os.environ or
        'APPVEYOR' in os.environ or
        'READTHEDOCS' in os.environ
    )
    is_travis = 'TRAVIS' in os.environ or 'TRAVIS_PYTHON_VERSION' in os.environ
    is_rtd = 'READTHEDOCS' in os.environ

    #if is_dev or (is_gui and add_vtk_qt):
        #try:
            #import vtk
            #vtk_version = '.'.join(vtk.VTK_VERSION.split('.'))
            #all_reqs['vtk'] = vtk_version
            #if vtk_version < '7.0.0':
                #print("vtk.VTK_VERSION = %r < '7.0.0'" % vtk.VTK_VERSION)
                #install_requires.append('vtk >= 7.0.0')
        #except ImportError:
            #install_requires.append('vtk >= 7.0.0')  # 8.x used

    install_requires = []

    if is_rtd:
        install_requires.append('numpy')
    else:
        version_check, required_version = vreqs['numpy']
        try:
            import numpy as np
            sver = np.lib.NumpyVersion(np.__version__)
            iver = int_version('numpy', sver.version)
            all_reqs['numpy'] = sver.version

            iversion_check = int_version('numpy', version_check)
            #srequired_version = int_version('numpy', required_version)
            #print('numpy %r %r' % (sver, iversion_check))
            if iver < iversion_check:
                print("numpy.__version__ = %r < %s" % (np.__version__, version_check))
                install_requires.append('numpy %s' % required_version)
                all_reqs['numpy'] = version_check
                install_requires.append('numpy %s' % required_version) # was 1.11; ,<1.13.0
        except ImportError:
            all_reqs['numpy'] = required_version
            install_requires.append('numpy %s' % required_version) # ,<1.13.0; 1.15.1 used

    if is_rtd:
        install_requires.append('scipy')
    else:
        version_check, required_version = vreqs['scipy']
        try:
            import scipy
            sver = scipy.version.short_version
            iver = int_version('scipy', sver)
            all_reqs['scipy'] = sver

            iversion_check = int_version('scipy', version_check)
            #srequired_version = int_version('scipy', required_version)
            #print('scipy %r %r' % (sver, iversion_check))
            #print(iver, iversion_check)
            if iver < iversion_check:
                print("scipy.version.short_version = %r < %r" % (
                    scipy.version.short_version, version_check))
                all_reqs['scipy'] = required_version
                install_requires.append('scipy %s' % required_version)
        except ImportError:
            install_requires.append('scipy %s' % required_version)  # 1.1.0 used

    try:
        import six
        iver = int_version('six', six.__version__)
        all_reqs['six'] = str_version(iver)
        if iver < [1, 9, 0]:
            print("six.__version__ = %r < '1.9.0'" % six.__version__)
            all_reqs['six'] = '>= 1.9.0'
            install_requires.append('six >= 1.9.0')
    except ImportError:
        install_requires.append('six >= 1.9.0')  # 1.12.0 used

    if is_gui:
        version_check, required_version = vreqs['matplotlib']
        try:
            import matplotlib
            iver = int_version('matplotlib', matplotlib.__version__)
            all_reqs['matplotlib'] = str_version(iver)
            iversion_check = int_version('matplotlib', version_check)
            if iver < iversion_check:
                print("matplotlib.__version__ = %r < %r" % (matplotlib.__version__, version_check))
                    #matplotlib.__version__, str_version(iversion_check)))
                all_reqs['matplotlib'] = required_version # '>= 2.1.2'
                install_requires.append('matplotlib %s' % required_version)
        except ImportError:
            install_requires.append('matplotlib %s' % required_version)  # was 2.1.2; 2.2.3 used


    try:
        import cpylog
        iver = int_version('cpylog', cpylog.__version__)
        all_reqs['cpylog'] = str_version(iver)
        if (PY2 or PY36) and not([1, 0, 2] <= iver < [1, 3]):
            print("cpylog.__version__ = %r '> 1.0.2, < 1.3'" % cpylog.__version__)
            all_reqs['cpylog'] = '>= 1.0.2, <1.3'
            install_requires.append('cpylog >= 1.0.2,<1.3')
        elif [1, 0, 2] <= iver:
            print("cpylog.__version__ = %r != '1.0.2'" % cpylog.__version__)
            all_reqs['cpylog'] = '>= 1.0.2'
            install_requires.append('cpylog >= 1.0.2')
    except ImportError:
        # 1.0.2 used
        if PY2 or PY36:
            install_requires.append('cpylog >= 1.0.2,<1.3')
        else:
            install_requires.append('cpylog >= 1.0.2')

    try:
        import docopt
        iver = int_version('docopt', docopt.__version__)
        all_reqs['docopt'] = str_version(iver)
        if iver != [0, 6, 2]:
            print("docopt.__version__ = %r != '0.6.2'" % docopt.__version__)
            all_reqs['docopt'] = '== 0.6.2'
            install_requires.append('docopt == 0.6.2')
    except ImportError:
        install_requires.append('docopt == 0.6.2')  # 0.6.2 used

    if is_rtd:
        pass
    elif is_gui:
        try:
            import qtpy
            iver = int_version('qtpy', qtpy.__version__)
            all_reqs['qtpy'] = str_version(iver)
            if iver < [1, 4, 0]:
                print("qtpy.__version__ = %r < '1.4.0'" % qtpy.__version__)
                all_reqs['qtpy'] = '>= 1.4.0'
                install_requires.append('qtpy >= 1.4.0')
        except ImportError:
            install_requires.append('qtpy >= 1.4.0')  # 1.5.0 used

    if is_rtd:
        pass
    try:
        import pandas
        iver = int_version('pandas', pandas.__version__)
        all_reqs['pandas'] = str_version(iver)
        if not([0, 20] <= iver < [0, 25]):
            print("pandas.__version__ = %r >= '0.20, <0.25'" % pandas.__version__)
            all_reqs['pandas'] = 'pandas > 0.20, <0.25'
            install_requires.append('pandas > 0.20, <0.25')
    except ImportError:
        pass
        #install_requires.append('pandas > 0.20, <0.25')

    if is_rtd:
        pass
    elif is_gui:
        try:
            import PIL
            iver = int_version('pillow', PIL.__version__)
            all_reqs['pillow'] = str_version(iver)
            if PY2 and not ([1, 4, 0] and iver < [7, 0]):
                print("pillow.__version__ = %r '>= 1.4.0, < 7.0'" % PIL.__version__)
                all_reqs['pillow'] = 'pillow >=1.4.0, < 7.0'
                install_requires.append('pillow >=1.4.0, < 7.0')
            elif iver < [1, 4, 0]:
                print("pillow.__version__ = %r > '1.4.0'" % PIL.__version__)
                all_reqs['pillow'] = 'pillow >=1.4.0'
                install_requires.append('pillow >= 1.4.0')
        except ImportError:
            #print('missing pillow...')
            if PY2:
                install_requires.append('pillow >= 1.4.0, < 7.0')  # 1.5.0 used
            else:
                install_requires.append('pillow >= 1.4.0')  # 1.5.0 used

        try:
            import imageio
            if imageio.__version__ < '2.2.0':
                print("imageio.version = %r < '2.2.0'" % imageio.__version__)
                all_reqs['imageio'] = '>= 2.2.0'
                install_requires.append('imageio >= 2.2.0')
            else:
                all_reqs['imageio'] = imageio.__version__
        except ImportError:
            install_requires.append('imageio >= 2.2.0')  # 2.6.1 used

    if PY2:
        try:
            import typing
        except ImportError:
            # PY2
            all_reqs['typing'] = '>= 3.6.4'
            install_requires.append('typing >= 3.6.4')  # 3.6.6 used

        try:
            import pathlib2
        except ImportError:
            all_reqs['pathlib2'] = '>= 2.3.0'
            install_requires.append('pathlib2 >= 2.3.0')  # 2.3.2 used

        #try:
            #import scandir
            #iver = int_version(scandir.__version__)
            #all_reqs['scandir'] = str_version(iver)
            #if sver < [1, 7, 0]:
                #print("scandir.__version__ = %r < '1.7.0'" % scandir.__version__)
                #all_reqs['scandir'] = '>= 1.7.0'
                #install_requires.append('scandir >= 1.7.0')
        #except ImportError:
            #install_requires.append('scandir >= 1.7.0')  # 1.9.0 used



    #is_windows = 'nt' in os.name
    if is_travis:
        #install_requires.append('coverage')
        #install_requires.append('python-coveralls')
        install_requires.append('codecov')

    #print(all_reqs)
    print('install_requires = %s' %  str(install_requires))
    return all_reqs, install_requires
